Sources/TensorFlow/Epochs/NonuniformTrainingEpochs.swift

```diff
@@ -0,0 +1,179 @@
+  }
+
+  /// The type of each epoch, a collection of batches of samples.
+  public typealias Element = Slices<
+    LazilySelected<Samples, Array<Samples.Index>.SubSequence>
+  >
+
+  /// Returns the next epoch in sequence.
+  public func next() -> Element? {
+    let remainder = sampleOrder.count % batchSize
+
+    sampleOrder.withUnsafeMutableBufferPointer { order in
+      // TODO: use a parallel shuffle like mergeshuffle
+      // (http://ceur-ws.org/Vol-2113/paper3.pdf)
+      order.shuffle(using: &entropy)
+
+      // The indices of samples used in this epoch
+      var epochSampleOrder = order.dropLast(remainder)
+
+      // The index in order of the largest sample.
+      let leader = epochSampleOrder.indices.max {
+        areInAscendingSizeOrder(
+          samples[epochSampleOrder[$0]], samples[epochSampleOrder[$1]])
+      }!
+
+      // The last position in epochSamples that will end up in the first batch.
+      let lastOfFirstBatch = epochSampleOrder.index(atOffset: batchSize - 1)
+      if leader > lastOfFirstBatch {
+        epochSampleOrder[lastOfFirstBatch...leader]
+          .rotate(shiftingToStart: leader)
+      }
+
+      // The regions of usedOrder to be sorted by descending batch size
+      let megabatches =
+        epochSampleOrder
+        .inBatches(of: batchSize * batchesPerSort)
+
+      for var megabatch in megabatches {
+        // TODO: fully sorting is overkill; we should use introselect here.
+        // Also, parallelize.
+        megabatch.sort { areInAscendingSizeOrder(samples[$1], samples[$0]) }
+      }
+    }
+    return samples.sampled(at: sampleOrder.dropLast(remainder))
+      .inBatches(of: batchSize)
+  }
+}
+
+extension NonuniformTrainingEpochs
+where Entropy == SystemRandomNumberGenerator {
+  /// Creates an instance drawing samples from `samples` into batches of size
+  /// `batchSize`.
+  ///
+  /// - Parameters:
+  ///   - batchesPerSort: the number of batches across which to group sample
+  ///     sizes similarly, or `nil` to indicate that the implementation should
+  ///     choose a number. Choosing too high can destroy the effects of sample
+  ///     shuffling in many training schemes, leading to poor results.  Choosing
+  ///     too low will reduce the similarity of sizes in a given batch, leading
+  ///     to inefficiency.
+  ///   - areInAscendingSizeOrder: a predicate that returns `true` iff the size
+  ///     of the first parameter is less than that of the second.
+  public convenience init(
+    samples: Samples,
+    batchSize: Int,
+    batchesPerSort: Int? = nil,
+    areInAscendingSizeOrder:
+      @escaping (Samples.Element, Samples.Element) -> Bool
+  ) {
+    self.init(
+      samples: samples,
+      batchSize: batchSize,
+      entropy: SystemRandomNumberGenerator(),
+      batchesPerSort: batchesPerSort,
+      areInAscendingSizeOrder: areInAscendingSizeOrder
+    )
+  }
+}
+
+/// Build batches for inference drawing samples from `samples` into batches of 
```

Comment:
Writing an extension on `Slices where Base == LazilySelected<Samples, [Samples.Index]>` (well now `Slices where Base == Sampling<Samples, [Samples.Index]>` does not work out of the box since it has `Samples` as a generic.

Reply:
Yeah we'd need a `_SamplingProtocol` to work around the language limitation.

Reply:
The other option of course is to actually make it a type that forwards all of its implementations to the thing it wraps.  I think consistency of this thing for the user is still an important surface API point.